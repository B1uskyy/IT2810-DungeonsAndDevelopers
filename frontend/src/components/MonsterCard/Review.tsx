import { Slider, Box } from '@mui/material';
import { GiDaemonSkull, GiGoblinHead, GiRoundShield, GiSpikedDragonHead } from 'react-icons/gi';
import { LuSwords } from 'react-icons/lu';

type ReviewProps = {
  review: {
    user: string;
    difficulty: number;
    description: string;
  };
};

const marks = [
  { value: 10, label: <LuSwords size={20} /> },
  { value: 30, label: <GiRoundShield size={20} /> },
  { value: 50, label: <GiGoblinHead size={20} /> },
  { value: 70, label: <GiSpikedDragonHead size={20} /> },
  { value: 90, label: <GiDaemonSkull size={20} /> },
];

const Review = ({ review }: ReviewProps) => {
  const { user, difficulty, description } = review;

  return (
    <Box className="border-[1px] border-white p-4 rounded mb-4 gap-6 flex flex-col">
      <h4 className="sub-header">{user}</h4>
      <Slider
        value={difficulty}
        marks={marks}
        disabled
        sx={{
          '& .MuiSlider-markLabel': {
            color: 'white',
            fontFamily: 'MedievalSharp',
            fontSize: '1.5rem',
          },

          '& .MuiSlider-mark': {
            color: 'white',
            width: 5,
            height: 5,
            borderRadius: '50%',
            transform: 'translateX(-50%) translateY(-50%)',
          },

          '& .MuiSlider-thumb': {
            display: 'none',
          },
          '& .MuiSlider-track': {
            color: '#DB3232',
            height: 10,
          },
          '& .MuiSlider-rail': {
            color: '#DB3232',
            height: 10,
          },
        }}
      />
      <p className="text">Difficulty: {difficulty}</p>
      <p className="text">{description}</p>
    </Box>
  );
};

export default Review;