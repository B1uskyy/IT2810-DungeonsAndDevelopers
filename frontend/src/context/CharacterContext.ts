import AbilityScoreCardProps from '../interfaces/AbilityScoreProps.ts';
import ClassData from '../interfaces/ClassProps.ts';
import RaceData from '../interfaces/RaceProps.ts';
import { Equipment } from '../interfaces/EquipmentProps.ts';
import { createContext, useContext } from 'react';

interface CharacterContextType {
  stateAbilities: AbilityScoreCardProps[];
  userAbilityScores: Map<string, number>;
  updateAbilityScores: (newValue: number, updatedAbilityName: string) => Promise<void>;
  classes: ClassData[];
  updateClass: (classId: string) => Promise<void>;
  races: RaceData[];
  updateRace: (raceId: string) => Promise<void>;
  addToEquipments: (equipment: Equipment) => void;
  removeFromEquipments: (equipment: Equipment) => void;
  removeAllEquipments: () => void;
  loadingStates: LoadingStates;
}

interface LoadingStates {
  abilityScoresLoading: boolean;
  classLoading: boolean;
  raceLoading: boolean;
  equipmentsLoading: boolean;
}

export const CharacterContext = createContext<CharacterContextType>({
  stateAbilities: [],
  userAbilityScores: new Map<string, number>(),
  updateAbilityScores: async () => Promise.resolve(),
  classes: [],
  updateClass: async () => Promise.resolve(),
  races: [],
  updateRace: async () => Promise.resolve(),
  addToEquipments: () => {},
  removeFromEquipments: () => {},
  removeAllEquipments: () => {},
  loadingStates: {
    equipmentsLoading: false,
    classLoading: false,
    abilityScoresLoading: false,
    raceLoading: false,
  },
});

export const useCharacterContext = () => {
  const context = useContext(CharacterContext);
  if (!context) {
    throw new Error('useCharacterContext must be used within a CharacterProvider');
  }
  return context;
};
