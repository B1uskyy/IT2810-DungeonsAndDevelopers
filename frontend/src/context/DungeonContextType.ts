import { createContext } from 'react';
import { MonsterCardProps } from '../interfaces/MonsterCardProps';

export interface DungeonContextType {
  dungeonMonsters: MonsterCardProps[];
  toggleDungeon: (monster: MonsterCardProps) => void;
  isInDungeon: (monsterIndex: string) => boolean;
}

export const DungeonContext = createContext<DungeonContextType>({
  dungeonMonsters: [],
  toggleDungeon: () => {},
  isInDungeon: () => false,
});
