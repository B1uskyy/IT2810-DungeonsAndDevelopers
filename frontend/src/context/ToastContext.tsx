import React, { createContext } from 'react';

/**
 * ToastContext
 *
 * Provides a context for managing toast notifications.
 *
 * Features:
 * - Displays toast notifications with customizable messages, types, and durations.
 * - Supports undo actions and manual toast closure.
 * - Ensures a consistent interface for triggering toast notifications across the application.
 *
 * Interfaces:
 * - `ToastProps`: Defines the properties of a toast notification.
 * - `ToastContextType`: Describes the methods and state exposed by the ToastContext.
 */

export interface ToastProps {
  message: string;
  type: 'success' | 'error' | 'info' | 'warning';
  duration?: number;
  undoAction?: () => void;
  closeToast?: () => void;
}

export interface ToastContextType {
  showToast: (props: ToastProps) => void;
  toastId: React.MutableRefObject<string | null>;
}

export const ToastContext = createContext<ToastContextType | undefined>(undefined);
