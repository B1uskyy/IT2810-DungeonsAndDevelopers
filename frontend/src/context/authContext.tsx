import { createContext, useReducer } from 'react';



const initialState = {
  user: ""
};

interface LoginState{
  user: string
}

if (localStorage.getItem('token')) {
  initialState.user = JSON.parse(localStorage.getItem('token')!);
}

const AuthContext = createContext({
  user: "",
  login: (useData) => {
  },
});

function authReducer( state: LoginState, action: any ) {
  switch (action.type) {
    case 'LOGIN':
      return {
        ...state,
        user: action.payload,
      };
    default:
      return state;
  }
}

function AuthProvider() {
  const [state, dispatch] = useReducer(authReducer, { user: "" })

  const login = (useData: { token: string; }) => {
    localStorage.setItem('token', useData.token);

    dispatch({ type: 'LOGIN', payload: useData})}



  return(
    <>
      <AuthContext.Provider
      value={{user: state.user, login}}
      />
    </>
  )
}
export {AuthContext, AuthProvider}