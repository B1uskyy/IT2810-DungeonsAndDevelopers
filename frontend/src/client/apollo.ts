import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
