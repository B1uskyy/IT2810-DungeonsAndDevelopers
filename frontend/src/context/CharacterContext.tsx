import { makeVar, useMutation, useQuery, useReactiveVar } from '@apollo/client';
import { createContext, ReactNode, useCallback, useEffect, useMemo } from 'react';

import { UPDATE_ABILITY_SCORES, UPDATE_USER_CLASS, UPDATE_USER_RACE } from '../graphql/updateUserQueries.ts';
import { GET_ARRAY_SCORES, GET_USER_CLASS, GET_USER_RACE } from '../graphql/userQueries.ts';
import useAbilityScores from '../hooks/useAbilityScores.ts';
import useClasses from '../hooks/useClasses.ts';
import useRaces from '../hooks/useRaces.ts';
import { useToast } from '../hooks/useToast.ts';
import useUserEquipments from '../hooks/useUserEquipments';
import AbilityScoreCardProps from '../interfaces/AbilityScoreProps.ts';
import ClassData from '../interfaces/ClassProps.ts';
import { Equipment } from '../interfaces/EquipmentProps';
import RaceData from '../interfaces/RaceProps.ts';
import { abilitiesVar } from '../pages/mainPages/myCharacterPage.tsx';
import { classVar } from '../pages/subPages/classPage.tsx';
import { raceVar } from '../pages/subPages/racePage.tsx';
import { notifyScoreChanges } from '../utils/abilityScoreMapping.ts';

interface CharacterContextType {
  stateAbilities: AbilityScoreCardProps[];
  userAbilityScores: Map<string, number>;
  updateAbilityScores: (newValue: number, updatedAbilityName: string) => Promise<void>;
  classes: ClassData[];
  updateClass: (classId: string) => Promise<void>;
  races: RaceData[];
  updateRace: (raceId: string) => Promise<void>;
  addToEquipments: (equipment: Equipment) => void;
  removeFromEquipments: (equipment: Equipment) => void;
  removeAllEquipments: () => void;
  loadingStates: LoadingStates;
}

interface LoadingStates {
  abilityScoresLoading: boolean;
  classLoading: boolean;
  raceLoading: boolean;
  equipmentsLoading: boolean;
}

interface CharacterProviderProps {
  children: ReactNode;
  userId: string;
}

interface UserClass {
  user: {
    id: string;
    class: {
      id: string;
      name: string;
      index: string;
      hit_die: number;
    };
  };
}

interface UserRace {
  user: {
    id: string;
    race: {
      id: string;
      name: string;
      speed: number;
      alignment: string;
      size: string;
      img: string;
    };
  };
}
interface UserAbilities {
  user: {
    id: string;
    abilityScores: {
      ability: {
        id: string;
        name: string;
      };
      score: number;
    };
  };
}
interface Ability {
  id: string;
  name: string;
}

interface AbilityScorePair {
  __typename: string;
  ability: Ability;
  score: number;
}

interface ArrayScores {
  getArrayScores: AbilityScorePair[];
}

type ArrayVar = {
  userId: string;
};

export const equipmentsVar = makeVar<Equipment[]>([]);

export const CharacterContext = createContext<CharacterContextType>({
  stateAbilities: [],
  userAbilityScores: new Map<string, number>(),
  updateAbilityScores: async () => Promise.resolve(),
  classes: [],
  updateClass: async () => Promise.resolve(),
  races: [],
  updateRace: async () => Promise.resolve(),
  addToEquipments: () => {},
  removeFromEquipments: () => {},
  removeAllEquipments: () => {},
  loadingStates: {
    equipmentsLoading: false,
    classLoading: false,
    abilityScoresLoading: false,
    raceLoading: false,
  },
});

export const CharacterProvider = ({ children, userId }: CharacterProviderProps) => {
  const { showToast } = useToast();

  const currentEquipments = useReactiveVar(equipmentsVar);

  const { classes: fetchedClasses } = useClasses(1, 12);
  const { races: fetchedRaces } = useRaces(1, 12);
  const { abilities: dataAbilities } = useAbilityScores(1, 6);

  const { data: scoreData, loading: abilityScoresLoading } = useQuery<ArrayScores, ArrayVar>(GET_ARRAY_SCORES, {
    variables: { userId },
    skip: !userId,
    fetchPolicy: 'cache-and-network',
  });

  const { data: userClassData, loading: classLoading } = useQuery(GET_USER_CLASS, {
    variables: { userId },
    skip: !userId,
    fetchPolicy: 'cache-and-network',
  });

  const { data: userRaceData, loading: raceLoading } = useQuery(GET_USER_RACE, {
    variables: { userId },
    skip: !userId,
    fetchPolicy: 'cache-and-network',
  });

  const abilityScores = useReactiveVar(abilitiesVar);

  const {
    userEquipments,
    removeFromEquipmentsMutation,
    addToEquipmentsMutation,
    removeAllUserEquipmentsMutation,
    loading: equipmentsLoading,
  } = useUserEquipments();

  const selectedClassId = userClassData?.user?.class?.id || '';
  const selectedRaceId = userRaceData?.user?.race?.id || '';

  useEffect(() => {
    equipmentsVar(userEquipments);
  }, [currentEquipments, userEquipments]);

  useEffect(() => {
    classVar(selectedClassId);
  }, [selectedClassId]);

  useEffect(() => {
    raceVar(selectedRaceId);
  }, [selectedRaceId]);

  useEffect(() => {
    if (scoreData?.getArrayScores) {
      const mappedScores = new Map(
        scoreData.getArrayScores.map((item: AbilityScorePair) => [item.ability.name, item.score])
      );
      abilitiesVar(mappedScores);
    }
  }, [scoreData]);

  const classes = useMemo(() => fetchedClasses || [], [fetchedClasses]);
  const races = useMemo(() => fetchedRaces || [], [fetchedRaces]);

  const checkUser = useCallback(
    (message: string) => {
      if (!userId) {
        showToast({
          message: `You must be logged in to change ${message}`,
          type: 'error',
          duration: 5000,
        });
        return false;
      }
      return true;
    },
    [showToast, userId]
  );

  const addToEquipments = async (equipment: Equipment) => {
    try {
      await addToEquipmentsMutation(equipment.id);
      equipmentsVar([...equipmentsVar(), equipment]);
    } catch (error) {
      console.error('Error adding equipment:', error);
    }
  };

  const removeFromEquipments = async (equipment: Equipment) => {
    try {
      await removeFromEquipmentsMutation(equipment.id);

      equipmentsVar(equipmentsVar().filter((equip) => equip.id !== equipment.id));
    } catch (error) {
      console.error('Error removing equipment:', error);
    }
  };

  const removeAllEquipments = async () => {
    try {
      await removeAllUserEquipmentsMutation();
      equipmentsVar([]);
    } catch (error) {
      console.error('Error removing all equipment:', error);
    }
  };

  const [updateAbilityScoresMutation] = useMutation(UPDATE_ABILITY_SCORES, {
    update: (cache, { data }) => {
      if (!data) return;

      const existing = cache.readQuery<UserAbilities>({
        query: GET_ARRAY_SCORES,
        variables: { userId },
      });

      if (existing?.user) {
        cache.writeQuery({
          query: GET_ARRAY_SCORES,
          variables: { userId },
          data: {
            user: {
              __typename: 'User',
              id: userId,
              abilityScores: {
                ...existing?.user.abilityScores.ability,
                score: data.user.score,
              },
            },
          },
        });
      }
    },
    onError: (error) => {
      console.error('Failed to update ability scores: ', error);

      showToast({
        message: `Failed to update ability scores: ${error.message}`,
        type: 'error',
        duration: 3000,
      });
    },
  });

  const [updateUserClassMutation] = useMutation(UPDATE_USER_CLASS, {
    update: (cache, { data }) => {
      if (!data) return;
      const existing = cache.readQuery<UserClass>({
        query: GET_USER_CLASS,
        variables: { userId },
      });
      if (existing?.user) {
        cache.writeQuery({
          query: GET_USER_CLASS,
          variables: { userId },
          data: {
            user: {
              __typename: 'User',
              id: userId,
              class: {
                ...data.updateUserClass.class,
              },
            },
          },
        });
      }
    },
  });

  const [updateUserRaceMutation] = useMutation(UPDATE_USER_RACE, {
    update: (cache, { data }) => {
      if (!data) return;
      const existing = cache.readQuery<UserRace>({
        query: GET_USER_RACE,
        variables: { userId },
      });
      if (existing?.user) {
        cache.writeQuery({
          query: GET_USER_RACE,
          variables: { userId },
          data: {
            user: {
              __typename: 'User',
              id: userId,
              race: {
                ...data.updateUserRace.race,
              },
            },
          },
        });
      }
    },
  });

  const updateAbilityScores = useCallback(
    async (newValue: number, name: string) => {
      if (!checkUser('Abilities')) return;

      const updatedScores = new Map(abilitiesVar());
      updatedScores.set(name, newValue);

      try {
        const scoresArray = Array.from(updatedScores.values());

        await updateAbilityScoresMutation({
          variables: { userId, scores: scoresArray },
        });

        abilitiesVar(updatedScores);
        notifyScoreChanges(updatedScores, abilitiesVar(), abilitiesVar, showToast);
      } catch (error) {
        console.error(error);
      }
    },
    [checkUser, showToast, updateAbilityScoresMutation, userId]
  );

  const updateClass = async (classId: string) => {
    if (!checkUser('Class')) return;

    try {
      const response = await updateUserClassMutation({ variables: { userId, classId } });
      classVar(classId);
      const updatedClass = response.data.updateUserClass.class;
      showToast({
        message: `Class changed to ${updatedClass.name}`,
        type: 'success',
        duration: 3000,
      });
    } catch (error) {
      showToast({
        message: `Failed to update class: ${error}`,
        type: 'error',
        duration: 3000,
      });
    }
  };

  const updateRace = async (raceId: string) => {
    if (!checkUser('Race')) return;

    try {
      const response = await updateUserRaceMutation({ variables: { userId, raceId } });
      raceVar(raceId);
      const updatedRace = response.data.updateUserRace.race;

      showToast({
        message: `Race changed to ${updatedRace.name}`,
        type: 'success',
        duration: 3000,
      });
    } catch (error) {
      console.error('Error updating race:', error);
      showToast({
        message: 'Failed to update race. Please try again.',
        type: 'error',
        duration: 3000,
      });
    }
  };

  const loadingStates = {
    abilityScoresLoading,
    classLoading,
    raceLoading,
    equipmentsLoading,
  };

  return (
    <CharacterContext.Provider
      value={{
        stateAbilities: dataAbilities,
        userAbilityScores: abilityScores,
        updateAbilityScores,
        classes,
        updateClass,
        races,
        updateRace,
        addToEquipments,
        removeFromEquipments,
        removeAllEquipments,
        loadingStates,
      }}
    >
      {children}
    </CharacterContext.Provider>
  );
};
