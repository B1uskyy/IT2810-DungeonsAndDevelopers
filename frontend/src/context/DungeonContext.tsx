import { createContext, ReactNode, useCallback } from 'react';

import { makeVar, useReactiveVar } from '@apollo/client';
import useDungeon from '../hooks/useDungeon.ts';
import { useToast } from '../hooks/useToast.ts';
import { MonsterCardProps } from '../interfaces/MonsterCardProps.ts';

interface DungeonContextType {
  dungeonMonsters: MonsterCardProps[];
  dungeonName: string;
  toggleDungeon: (monster: MonsterCardProps) => void;
  isInDungeon: (monsterIndex: string) => boolean;
  updateDungeonName: (newName: string) => void;
}

interface DungeonProviderProps {
  children: ReactNode;
  userId: string;
}

export const dungeonMonstersVar = makeVar<MonsterCardProps[]>([]);

export const DungeonContext = createContext<DungeonContextType>({
  dungeonMonsters: [],
  dungeonName: '',
  toggleDungeon: () => {},
  isInDungeon: () => false,
  updateDungeonName: () => {},
});

export const DungeonProvider = ({ children, userId }: DungeonProviderProps) => {
  const maxFavorites = 6;
  const { showToast } = useToast();

  const { dungeonMonsters, dungeonName, toggleFavorite, toggleDungeonName } = useDungeon();

  const currentDungeonMonsters = useReactiveVar(dungeonMonstersVar);

  const isInDungeon = useCallback(
    (monsterId: string) => {
      return currentDungeonMonsters.some((favMonster) => favMonster.id === monsterId);
    },
    [currentDungeonMonsters]
  );
  const toggleDungeon = async (monster: MonsterCardProps) => {
    if (!userId) {
      showToast({ message: 'You must be logged in to add monsters to dungeon', type: 'error', duration: 3000 });
      return;
    }

    if (!isInDungeon(monster.id) && dungeonMonsters.length >= maxFavorites) {
      showToast({ message: 'You can only add 6 monsters to your dungeon', type: 'warning', duration: 3000 });
      return;
    }

    const wasInDungeon = isInDungeon(monster.id);

    try {
      await toggleFavorite(monster);
      if (wasInDungeon) {
        showToast({
          message: `${monster.name} removed from dungeon`,
          type: 'info',
          duration: 5000,
          undoAction: () => undoRemoveMonster(monster),
        });
      } else {
        showToast({
          message: `${monster.name} added to dungeon`,
          type: 'success',
          duration: 3000,
        });
      }
    } catch (error) {
      console.error('Error in toggleDungeon:', error);
    }
  };

  const updateDungeonName = async (newName: string) => {
    try {
      await toggleDungeonName(newName);
      showToast({
        message: `Dungeon name updated to "${newName}"`,
        type: 'success',
        duration: 3000,
      });
    } catch (error) {
      console.error('Error updating dungeon name:', error);
      showToast({
        message: 'Failed to update dungeon name. Please try again.',
        type: 'error',
        duration: 3000,
      });
    }
  };

  const undoRemoveMonster = async (monster: MonsterCardProps) => {
    try {
      await toggleFavorite(monster);
      showToast({
        message: `${monster.name} restored to dungeon`,
        type: 'success',
        duration: 3000,
      });
      console.log('After undo: ', dungeonMonstersVar());
    } catch (error) {
      console.error('Error restoring monster to dungeon:', error);
      showToast({
        message: 'Failed to restore monster. Please try again.',
        type: 'error',
        duration: 3000,
      });
    }
  };

  return (
    <DungeonContext.Provider
      value={{
        dungeonMonsters: currentDungeonMonsters,
        dungeonName,
        toggleDungeon,
        isInDungeon,
        updateDungeonName,
      }}
    >
      {children}
    </DungeonContext.Provider>
  );
};
