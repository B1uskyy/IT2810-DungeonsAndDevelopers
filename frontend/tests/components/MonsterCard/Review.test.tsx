import { MockedProvider } from '@apollo/client/testing';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Review from '../../../src/components/MonsterCard/Review.tsx';
import { AuthContext } from '../../../src/context/AuthContext';
import { DELETE_REVIEW, GET_MONSTER_REVIEWS, UPDATE_REVIEW } from '../../../src/graphql/queries.ts';
import { ReviewType } from '../../../src/interfaces/ReviewProps';

// Mock the useToast hook
vi.mock('../../../src/hooks/useToast', () => ({
  useToast: () => ({
    showToast: vi.fn(),
  }),
}));

interface MockSliderProps {
  id: string;
  name: string;
  min: number;
  max: number;
  value: number;
  'data-testid': string;
  onChange: (event: Event | React.ChangeEvent<HTMLInputElement>, newValue: number) => void;
  disabled?: boolean;
}

// Mock the ReviewSlider component
vi.mock('../../../src/components/MonsterCard/ReviewSlider', () => ({
  default: ({ value, onChange, disabled = false }: MockSliderProps) => (
    <input
      type="range"
      data-testid="review-slider"
      value={value}
      onChange={(e) => {
        if (onChange && !disabled) {
          const newValue = Number(e.target.value);
          onChange(e, newValue);
        }
      }}
      min={0}
      max={100}
      step={10}
      disabled={disabled}
    />
  ),
}));

const mockReview: ReviewType = {
  id: 'review-123',
  user: { id: 'user-123', userName: 'Test User' },
  difficulty: 50,
  description: 'This is a test review.',
  createdAt: '2024-11-08T12:00:00Z',
};

const mockMonsterId = 'monster-456';
const mockMonsterName = 'Test Monster';

const mockUserId = 'user-123';
const mockToken = 'mock-token';
const mockUsername = 'Test User';
const mockLogin = vi.fn();
const mockLogout = vi.fn();

const mocks = [
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockMonsterId },
    },
    result: {
      data: {
        monster: {
          reviews: [mockReview],
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_REVIEW,
      variables: {
        monsterId: mockMonsterId,
        reviewId: mockReview.id,
        review: {
          user: mockUserId,
          difficulty: 70,
          description: 'This is an updated review',
        },
      },
    },
    result: {
      data: {
        updateReview: {
          id: mockReview.id,
          user: { id: mockUserId, userName: 'Test User' },
          difficulty: 70,
          description: 'This is an updated review',
          createdAt: mockReview.createdAt,
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockMonsterId },
    },
    result: {
      data: {
        monster: {
          reviews: [
            {
              id: mockReview.id,
              user: { id: mockUserId, userName: 'Test User' },
              difficulty: 70,
              description: 'This is an updated review',
              createdAt: mockReview.createdAt,
            },
          ],
        },
      },
    },
  },
  {
    request: {
      query: DELETE_REVIEW,
      variables: { monsterId: mockMonsterId, reviewId: mockReview.id },
    },
    result: {
      data: {
        deleteReview: {
          reviews: [],
        },
      },
    },
  },
];

const renderComponent = () => {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AuthContext.Provider
        value={{
          userId: mockUserId,
          userName: mockUsername,
          logout: mockLogout,
          login: mockLogin,
          token: mockToken,
        }}
      >
        <Review review={mockReview} monsterId={mockMonsterId} monsterName={mockMonsterName} />
      </AuthContext.Provider>
    </MockedProvider>
  );
};

describe('Review Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders review content correctly', async () => {
    const { asFragment } = renderComponent();

    expect(screen.getByText(mockReview.user.userName)).toBeInTheDocument();
    expect(screen.getByText(`Difficulty: ${mockReview.difficulty}`)).toBeInTheDocument();
    expect(screen.getByText(mockReview.description)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('allows user to edit and save review', async () => {
    renderComponent();
    const user = userEvent.setup();
    await waitFor(() => {
      expect(screen.getByText(`Difficulty: ${mockReview.difficulty}`)).toBeInTheDocument();
    });

    await user.click(screen.getByRole('button', { name: 'Edit' }));

    const newDifficulty = 70;
    const newDescription = 'This is an updated review';

    const slider = screen.getByTestId('review-slider');
    fireEvent.change(slider, { target: { value: newDifficulty } });

    await waitFor(() => {
      expect(slider).toHaveValue(String(newDifficulty));
    });

    const textField = screen.getByRole('textbox');
    await user.clear(textField);
    await user.type(textField, newDescription);

    expect(textField).toHaveValue(newDescription);

    const saveButton = screen.getByRole('button', { name: 'Save' });
    await user.click(saveButton);

    await waitFor(() => {
      expect(saveButton).not.toBeInTheDocument();
    });
  });
});
