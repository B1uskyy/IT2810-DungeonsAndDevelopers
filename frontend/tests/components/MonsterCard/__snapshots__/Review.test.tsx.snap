// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Review Component > renders review content correctly 1`] = `
<DocumentFragment>
  <div
    class="relative rounded mb-4 gap-6 flex flex-col w-full MuiBox-root css-163ijbm"
  >
    <button
      aria-label="Delete"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1jguzmj-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
    <button
      aria-label="Edit"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1w6nj3n-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </button>
    <h4
      class="sub-header"
    >
      Test User
    </h4>
    <input
      data-testid="review-slider"
      disabled=""
      max="100"
      min="0"
      step="10"
      type="range"
      value="50"
    />
    <p
      class="text"
    >
      Difficulty: 50
    </p>
    <p
      class="text"
      style="word-wrap: break-word; overflow: hidden; text-overflow: ellipsis;"
    >
      This is a test review.
    </p>
  </div>
</DocumentFragment>
`;
